### PR TITLE
PAPI

### PR DESCRIPTION
**Pull request by G-Ragghianti**
Added support for some optional PAPI components.

### CODE REVIEW

**Comment by G-Ragghianti** [comment]
Let's see if I can summon @luszczek

File: var/spack/repos/builtin/packages/papi/package.py [hunk 1]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
     version('5.4.1', '9134a99219c79767a11463a76b0b01a2')
     version('5.3.0', '367961dd0ab426e5ae367c2713924ffb')
 
+    variant('example', default=True, description='')
+    variant('cuda', default=False, description='')
+    variant('nvml', default=False, description='')
+    variant('infiniband', default=False, description='')
+    variant('powercap', default=False, description='')
+    variant('rapl', default=False, description='')
+    variant('lmsensors', default=False, description='')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: All of these variants need descriptions

File: var/spack/repos/builtin/packages/papi/package.py [hunk 2]
```diff
@@ -20,8 +20,9 @@ class Papi(Package):
        components that expose performance measurement opportunites
```
Comment (luszczek) [inline var/spack/repos/builtin/packages/papi/package.py]: fix spelling of "opportunites"

File: var/spack/repos/builtin/packages/papi/package.py [hunk 3]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
...
+    depends_on('cuda', when='+cuda')
+    depends_on('cuda', when='+nvml')
+    depends_on('lm-sensors', when='+lmsensors')
+
     # Does not build with newer versions of gcc, see
     # https://bitbucket.org/icl/papi/issues/46/cannot-compile-on-arch-linux
     patch('https://bitbucket.org/icl/papi/commits/53de184a162b8a7edff48fed01a15980664e15b1/raw', sha256='64c57b3ad4026255238cc495df6abfacc41de391a0af497c27d0ac819444a1f8', when='@5.4.0:5.6.99%gcc@8:')
 
+    def setup_environment(self, spack_env, run_env):
+        if '+cuda' in self.spec or 'nvml' in self.spec:
```
Comment (luszczek) [inline var/spack/repos/builtin/packages/papi/package.py]: Is it `+nvml`?

File: var/spack/repos/builtin/packages/papi/package.py [hunk 4]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
...
+            spack_env.set('CUDA_DIR', self.spec['cuda'].prefix)
+            run_env.prepend_path('LD_LIBRARY_PATH',
+                                 join_path(self.spec['cuda'].prefix,
+                                           "extras/CUPTI/lib64"))
+
     def install(self, spec, prefix):
+        with working_dir("src/components/nvml"):
```
Comment (luszczek) [inline var/spack/repos/builtin/packages/papi/package.py]: consider swapping "with" and "if" to make variant more visible

File: var/spack/repos/builtin/packages/papi/package.py [hunk 5]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
...
+            if '+nvml' in spec:
+                configure_args = [
+                    "--with-nvml-incdir=%s/include" % spec['cuda'].prefix,
+                    "--with-nvml-libdir=%s/lib64/stubs" % spec['cuda'].prefix,
+                    "--with-cuda-dir=%s" % spec['cuda'].prefix]
+                configure(*configure_args)
+        with working_dir("src/components/lmsensors"):
```
Comment (luszczek) [inline var/spack/repos/builtin/packages/papi/package.py]: consider swapping "with" and "if" to make variant more visible

File: var/spack/repos/builtin/packages/papi/package.py [hunk 6]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
...
+            if '+lmsensors' in spec:
+                configure_args = [
+                    "--with-sensors_incdir=%s/include/sensors" %
+                    spec['lm-sensors'].prefix,
+                    "--with-sensors_libdir=%s/lib64" %
+                    spec['lm-sensors'].prefix]
+                configure(*configure_args)
         with working_dir("src"):
 
             configure_args = ["--prefix=%s" % prefix]
```
Comment (luszczek) [inline var/spack/repos/builtin/packages/papi/package.py]: consider `mpi` variant in the future

File: var/spack/repos/builtin/packages/papi/package.py (hunk 1 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/papi/package.py]: Thanks.  Added descriptions.

File: var/spack/repos/builtin/packages/papi/package.py (hunk 6 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/papi/package.py]: Yes, this is the first set of variants.  PAPI will need a variant for most of the optional components.

Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/papi/package.py]: What would the MPI variant do?

File: var/spack/repos/builtin/packages/papi/package.py [hunk 7]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
...
+    depends_on('cuda', when='+cuda')
+    depends_on('cuda', when='+nvml')
+    depends_on('lm-sensors', when='+lmsensors')
+
     # Does not build with newer versions of gcc, see
     # https://bitbucket.org/icl/papi/issues/46/cannot-compile-on-arch-linux
     patch('https://bitbucket.org/icl/papi/commits/53de184a162b8a7edff48fed01a15980664e15b1/raw', sha256='64c57b3ad4026255238cc495df6abfacc41de391a0af497c27d0ac819444a1f8', when='@5.4.0:5.6.99%gcc@8:')
 
+    def setup_environment(self, spack_env, run_env):
+        if '+cuda' in self.spec or '+nvml' in self.spec:
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: You could also check `if '^cuda' in self.spec`

File: var/spack/repos/builtin/packages/papi/package.py [hunk 8]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
...
+            spack_env.set('CUDA_DIR', self.spec['cuda'].prefix)
+            run_env.prepend_path('LD_LIBRARY_PATH',
+                                 join_path(self.spec['cuda'].prefix,
+                                           "extras/CUPTI/lib64"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: You can do `self.spec['cuda'].prefix.extras.CUPTI.lib64`

File: var/spack/repos/builtin/packages/papi/package.py [hunk 9]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
...
+
     def install(self, spec, prefix):
+        if '+nvml' in spec:
+            with working_dir("src/components/nvml"):
+                configure_args = [
+                    "--with-nvml-incdir=%s/include" % spec['cuda'].prefix,
+                    "--with-nvml-libdir=%s/lib64/stubs" % spec['cuda'].prefix,
+                    "--with-cuda-dir=%s" % spec['cuda'].prefix]
+                configure(*configure_args)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: Do these other `configure` calls not need `--prefix`?

File: var/spack/repos/builtin/packages/papi/package.py [hunk 10]
```diff
@@ -30,11 +31,48 @@ class Papi(Package):
...
+        if '+lmsensors' in spec:
+            with working_dir("src/components/lmsensors"):
+                configure_args = [
+                    "--with-sensors_incdir=%s/include/sensors" %
+                    spec['lm-sensors'].prefix,
+                    "--with-sensors_libdir=%s/lib64" %
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/papi/package.py]: Will the libraries always be in `lib64`? We've had a lot of problems with this in the past across different systems. I would use `spec['lm-sensors'].libs.directories[0]` instead.

File: var/spack/repos/builtin/packages/papi/package.py (hunk 9 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/papi/package.py]: Correct, the "--prefix" is derived from the top-level configure for the papi source.

File: var/spack/repos/builtin/packages/papi/package.py (hunk 7 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/papi/package.py]: Yes, that looks better.

File: var/spack/repos/builtin/packages/papi/package.py (hunk 10 above)
Comment (G-Ragghianti) [reply var/spack/repos/builtin/packages/papi/package.py]: OK, this required a modification to the lm-sensors package (included in the next set of changes).

File: var/spack/repos/builtin/packages/lm-sensors/package.py [hunk 11]
```diff
@@ -32,3 +32,10 @@ class LmSensors(MakefilePackage):
     def install(self, spec, prefix):
         make('install', 'PREFIX={0}'.format(prefix),
              'ETCDIR={0}/etc'.format(prefix))
+
+    @property
+    def libs(self):
+        shared = True if '+shared' in self.spec else False
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lm-sensors/package.py]: There is no shared variant for this package, so this line isn't necessary.

**Comment by G-Ragghianti** [comment]
I just found a problem with finding the correct libcuda.so.

**Pull request closed without merging**